src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,231 @@
+        emitterLogic.start()
+        sq.watchCompletion().onComplete { _ => emitterLogic.cancel() }(singleThreadEC)
+        emitterLogic
+      }
+      .initialTimeout(zooKeeperConnectionTimeout)
+      .concat(Source.single(LeadershipState.Standby(None)))
+      .via(EnrichedFlow.dedup(LeadershipState.Standby(None)))
+      .map { e =>
+        e match {
+          case LeadershipState.ElectedAsLeader =>
+            logger.info("Leader won.")
+          case LeadershipState.Standby(currentLeader) =>
+            logger.info(s"Leader defeated. Current leader: ${currentLeader.getOrElse("-")}")
+        }
+        e
+      }
+  }
+
+  @SuppressWarnings(Array("CatchThrowable"))
+  private class CuratorEventEmitter(
+      singleThreadExecutor: ExecutorService,
+      clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+      zooKeeperLeaderPath: String,
+      hostPort: String,
+      sq: SourceQueueWithComplete[LeadershipState]) extends Cancellable {
+
+    val client = clientCloseable.closeable
+    private lazy val singleThreadEC: ExecutionContext = ExecutionContext.fromExecutor(singleThreadExecutor)
+    private lazy val leaderHostPortMetric: Timer = Metrics.timer(ServiceMetric, getClass, "current-leader-host-port")
+    private val curatorLeaderLatchPath = zooKeeperLeaderPath + "-curator"
+    private lazy val latch = new LeaderLatch(client, curatorLeaderLatchPath, hostPort)
+    @volatile private var isStarted = false
+    @volatile private var _isCancelled = false
+
+    /* It is important that we re-register our watch _BEFORE_ we get the current the current leader. In Zookeeper,
+     * watches are one-time use only.
+     *
+     * The timeline of events looks like this
+     *
+     * 1) We register a watch
+     * 2) We query the current leader
+     * 3) We receive an event (indicating child removal / addition)
+     * 4) Repeat step 1
+     *
+     * By re-registering the watch before querying the state, we will not miss out on the latest leader change.
+     */
+    def longPollLeaderChange(retries: Int = 0): Unit = Future {
+      try {
+        if (latch.getState == LeaderLatch.State.STARTED)
+          client.getChildren
+            .usingWatcher(new CuratorWatcher {
+              override def process(event: WatchedEvent): Unit =
+                if (!_isCancelled) longPollLeaderChange()
+            })
+            .forPath(curatorLeaderLatchPath)
+        emitLeader()
+      } catch {
+        case ex: KeeperException.NoNodeException if retries < 100 =>
+          // Wait for node to be created
+          logger.info("retrying")
+          Thread.sleep(retries * 10L)
+          longPollLeaderChange(retries + 1)
+        case ex: Throwable =>
+          sq.fail(ex)
+      }
+    }(singleThreadEC)
+
+    /**
+      * Emit current leader. Does not fail on connection error
+      */
+    private def emitLeader(): Unit = {
```

Comment:
return type ` Future[QueueOfferResult]`?

Reply:
Not sure if this is helpful. `emitLeader()` emits the event on the event stream. Seems straightforward to me?